daemon/daemon_main.go

```diff
@@ -1328,6 +1335,42 @@ func runDaemon() {
 		}()
 	}
 
+	// Migrating the ENI datapath must happen before the API is served to
+	// prevent endpoints from being created. It also must be before the health
+	// initialization logic which creates the health endpoint, for the same
+	// reasons as the API being served. We want to ensure that this migration
+	// logic runs before any endpoint creates.
+	if option.Config.IPAM == option.IPAMENI {
+		migrated, failed, partial := linuxrouting.NewMigrator(
+			&interfaceNumberGetter{},
+		).MigrateENIDatapath(option.Config.EgressMultiHomeIPRuleCompat)
+		switch {
+		case failed == -1:
+			// No need to handle this case specifically because it is handled
+			// in the call already.
+		case migrated >= 0 && failed > 0:
+			msg := fmt.Sprintf(
+				"Failed to migrate ENI datapath. %d endpoints were successfully migrated, "+
+					"%d were detected as previously attempted to be migrated, "+
+					"and %d failed to migrate completely. "+
+					"The original datapath is still in-place, however it is recommended to retry the migration.",
+				migrated, failed, partial)
+			log.Error(msg)
+
+		case migrated >= 0 && failed == 0:
+			msg := fmt.Sprintf(
+				"Migration of ENI datapath successful, %d endpoints were migrated and none failed.",
+				migrated)
+			if partial > 0 {
+				msg += fmt.Sprintf(
+					"Found %d partially migrated endpoints, which indicate a previously attempted failed migration. "+
+						"Consider setting --%s=%v and restarting Cilium to rollback or manually intervening.",
+					partial, option.EgressMultiHomeIPRuleCompat, !option.Config.EgressMultiHomeIPRuleCompat)
+			}
```

Comment:
I think the TL;DR of my commentary here is "Can we just delete the code that handles partial migration?"

Reply:
They're just random state leftover, so yes we can just remove that code. I think we previously wanted to handle them because we didn't quite understand / validate that regardless of failures in the migration, there will always be the new or original datapath in-place. Given that, this code becomes of little value and it doesn't matter to the end user that there are partially migrated rules; the endpoints' connectivity is still working anyway and at some point, we've already logged a msg about the underlying issue that caused them to not be deleted.